In [338]:
import numpy as np 
import pandas as pd

In [339]:
movies = pd.read_csv('tmdb_5000_movies.csv')
credits = pd.read_csv('tmdb_5000_credits.csv')
pd.set_option('display.max_column',None)

In [340]:
credits

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4798      9367                               El Mariachi   
4799     72766                                 Newlyweds   
4800    231617                 Signed, Sealed, Delivered   
4801    126186                          Shanghai Calling   
4802     25975                         My Date with Drew   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4798  [{"cast_id": 1, "character": "El Mariachi", "c...   
4799  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4800  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4801  [{"cast_id": 3, "character": "Sam", "credit_id...   
4802  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  
0     [{"credit_id": "52fe48009251416c750aca23", "de...  
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2     [{"credit_id": "54805967c3a36829b5002c41", "de...  
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  
...                                                 ...  
4798  [{"credit_id": "52fe44eec3a36847f80b280b", "de...  
4799  [{"credit_id": "52fe487dc3a368484e0fb013", "de...  
4800  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...  
4801  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...  
4802  [{"credit_id": "58ce021b9251415a390165d9", "de...  

[4803 rows x 4 columns]

In [341]:
movies

budget                                             genres  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...         ...                                                ...   
4798     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4799       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4800          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4801          0                                                 []   
4802          0                [{"id": 99, "name": "Documentary"}]   

                                               homepage      id  \
0                           http://www.avatarmovie.com/   19995   
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4798                                                NaN    9367   
4799                                                NaN   72766   
4800  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4801                        http://shanghaicalling.com/  126186   
4802                                                NaN   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4798  [{"id": 5616, "name": "united states\u2013mexi...                es   
4799                                                 []                en   
4800  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4801                                                 []                en   
4802  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4798                               El Mariachi   
4799                                 Newlyweds   
4800                 Signed, Sealed, Delivered   
4801                          Shanghai Calling   
4802                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4798  El Mariachi just wants to play his guitar and ...   14.269792   
4799  A newlywed couple's honeymoon is upended by th...    0.642552   
4800  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4801  When ambitious New York attorney Sam is sent t...    0.857008   
4802  Ever since the second grade when he first saw ... 

In [342]:
movies = pd.merge(movies,credits,how = 'inner',on='title')

In [343]:
movies

budget                                             genres  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...         ...                                                ...   
4804     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4805       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4806          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4807          0                                                 []   
4808          0                [{"id": 99, "name": "Documentary"}]   

                                               homepage      id  \
0                           http://www.avatarmovie.com/   19995   
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4804                                                NaN    9367   
4805                                                NaN   72766   
4806  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4807                        http://shanghaicalling.com/  126186   
4808                                                NaN   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4804  [{"id": 5616, "name": "united states\u2013mexi...                es   
4805                                                 []                en   
4806  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4807                                                 []                en   
4808  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4804  El Mariachi just wants to play his guitar and ...   14.269792   
4805  A newlywed couple's honeymoon is upended by th...    0.642552   
4806  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4807  When ambitious New York attorney Sam is sent t...    0.857008   
4808  Ever since the second grade when he first saw ... 

In [344]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

In [345]:
movies = movies[['id','title','overview','genres','keywords','cast','crew']]

In [346]:
movies.rename(columns={'id':'movie_id'},inplace=True)

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/3622227263.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies.rename(columns={'id':'movie_id'},inplace=True)


In [347]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4809 non-null   int64 
 1   title     4809 non-null   object
 2   overview  4806 non-null   object
 3   genres    4809 non-null   object
 4   keywords  4809 non-null   object
 5   cast      4809 non-null   object
 6   crew      4809 non-null   object
dtypes: int64(1), object(6)
memory usage: 263.1+ KB


In [348]:
movies.dropna(how='any',inplace=True)

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/233795588.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies.dropna(how='any',inplace=True)


In [349]:
movies.drop_duplicates(keep='first',inplace=True)

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/3608350129.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies.drop_duplicates(keep='first',inplace=True)


In [350]:
movies

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                                 genres  \
0     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...                                                 ...   
4804  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4805  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4806  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4807                                                 []   
4808                [{"id": 99, "name": "Documentary"}]   

                                               keywords  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4     [{"id": 818, "name": "based on novel"}, {"id":...   
...                                                 ...   
4804  [{"id": 5616, "name": "united states\u2013mexi...   
4805                                                 []   
4806  [{"id": 248, "name": "date"}, {"id": 699, "nam...   
4807                                                 []   
4808  [{"id": 1523, "name": "obsession"}, {"id": 224...   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4804  [{"cast_id": 1, "character": "El Mariachi", "c...   
4805  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4806  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4807  [{"cast_id": 3, "character": "Sam", "credit_id...   
4808  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  
0     [{"credit_id": "52fe48009251416c750aca23", "de...  
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2     [{"credit_id": "54805967c3a36829b5002c41", "de...  
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  
...                                                 ...  
4804  [{"credit_id": "52fe44eec3a36847f80b2

In [351]:
import ast
def funct(x):
    l2 = []
    l1 = ast.literal_eval(x)
    for i in l1:
        l2.append(i['name'])
    return l2

    
movies['genres'] = movies['genres'].apply(lambda x:funct(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/2005144596.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['genres'] = movies['genres'].apply(lambda x:funct(x))


In [352]:
movies['keywords'][0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [353]:
def funct(x):
    l1 = ast.literal_eval(x)
    l2 = []
    for i in l1:
        l2.append(i['name'])
    return l2


movies['keywords'] = movies['keywords'].apply(lambda x:funct(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/3085471167.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['keywords'] = movies['keywords'].apply(lambda x:funct(x))


In [354]:
movies['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [355]:
movies

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                             genres  \
0     [Action, Adventure, Fantasy, Science Fiction]   
1                      [Adventure, Fantasy, Action]   
2                        [Action, Adventure, Crime]   
3                  [Action, Crime, Drama, Thriller]   
4              [Action, Adventure, Science Fiction]   
...                                             ...   
4804                      [Action, Crime, Thriller]   
4805                              [Comedy, Romance]   
4806             [Comedy, Drama, Romance, TV Movie]   
4807                                             []   
4808                                  [Documentary]   

                                               keywords  \
0     [culture clash, future, space war, space colon...   
1     [ocean, drug abuse, exotic island, east india ...   
2     [spy, based on novel, secret agent, sequel, mi...   
3     [dc comics, crime fighter, terrorist, secret i...   
4     [based on novel, mars, medallion, space travel...   
...                                                 ...   
4804  [united states–mexico barrier, legs, arms, pap...   
4805                                                 []   
4806  [date, love at first sight, narration, investi...   
4807                                                 []   
4808          [obsession, camcorder, crush, dream girl]   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4804  [{"cast_id": 1, "character": "El Mariachi", "c...   
4805  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4806  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4807  [{"cast_id": 3, "character": "Sam", "credit_id...   
4808  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  
0     [{"credit_id": "52fe48009251416c750aca23", "de...  
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2     [{"credit_id": "54805967c3a36829b5002c41", "de...  
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  
...                                                 ...  
4804  [{"credit_id": "52fe44eec3a36847f80b280b", "de...  
4805  [{"credit_id": "52fe487dc3a

In [356]:
def funct(x):
    l1 = ast.literal_eval(x)
    l2 = []
    for i in l1:
        l2.append(i['name'])
    return l2
movies['cast'] = movies['cast'].apply(lambda x :funct(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/1505131338.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['cast'] = movies['cast'].apply(lambda x :funct(x))


In [357]:
type(movies['crew'][0])

str

In [358]:
def funct1(x):
    l2 = ast.literal_eval(x)
    l1 = []
    for i in l2:
        if i['job'] == 'Director':
            l1.append(i['name'])
            break
    return l1

movies['crew'] = movies['crew'].apply(lambda x :funct1(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/2808703099.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['crew'] = movies['crew'].apply(lambda x :funct1(x))


In [359]:
movies

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                             genres  \
0     [Action, Adventure, Fantasy, Science Fiction]   
1                      [Adventure, Fantasy, Action]   
2                        [Action, Adventure, Crime]   
3                  [Action, Crime, Drama, Thriller]   
4              [Action, Adventure, Science Fiction]   
...                                             ...   
4804                      [Action, Crime, Thriller]   
4805                              [Comedy, Romance]   
4806             [Comedy, Drama, Romance, TV Movie]   
4807                                             []   
4808                                  [Documentary]   

                                               keywords  \
0     [culture clash, future, space war, space colon...   
1     [ocean, drug abuse, exotic island, east india ...   
2     [spy, based on novel, secret agent, sequel, mi...   
3     [dc comics, crime fighter, terrorist, secret i...   
4     [based on novel, mars, medallion, space travel...   
...                                                 ...   
4804  [united states–mexico barrier, legs, arms, pap...   
4805                                                 []   
4806  [date, love at first sight, narration, investi...   
4807                                                 []   
4808          [obsession, camcorder, crush, dream girl]   

                                                   cast                 crew  
0     [Sam Worthington, Zoe Saldana, Sigourney Weave...      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley, ...     [Gore Verbinski]  
2     [Daniel Craig, Christoph Waltz, Léa Seydoux, R...         [Sam Mendes]  
3     [Christian Bale, Michael Caine, Gary Oldman, A...  [Christopher Nolan]  
4     [Taylor Kitsch, Lynn Collins, Samantha Morton,...     [Andrew Stanton]  
...                                                 ...                  ...  
4804  [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...   [Robert Rodriguez]  
4805  [Edward Burns, Kerry Bishé, Marsha Dietlein, C...       [Edward Burns]  
4806  [Eric Mabius, Kristin Booth, Crystal Lowe, Geo...        [Scott Smith]  
4807  [Daniel Henney, Eliza Coupe, Bill Paxton, Alan...        [Daniel Hsia]  
4808  [Drew Barrymore, Brian Herzlinger, Corey Feldm...   [Brian Herzlinger]  

[4806 rows x 7 columns]

In [360]:
def funct1(x):
    # x = ast.literal_eval(x)
    l1 = []
    for i in x:
        i = i.replace(' ','')
        l1.append(i)
    return l1
movies['cast'] = movies['cast'].apply(lambda x :funct1(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/824069408.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['cast'] = movies['cast'].apply(lambda x :funct1(x))


In [361]:
def funct(x):
    l1 = []
    for i in x:
        i = i.replace(' ','')
        l1.append(i)
    return l1
movies['crew'] = movies['crew'].apply(lambda x:funct(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/3762579540.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['crew'] = movies['crew'].apply(lambda x:funct(x))


In [362]:

def funct(x):
    w = x.split(' ')
    return w 
movies['overview'] = movies['overview'].apply(lambda x :funct(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/1986395503.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['overview'] = movies['overview'].apply(lambda x :funct(x))


In [363]:
movies

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     [In, the, 22nd, century,, a, paraplegic, Marin...   
1     [Captain, Barbossa,, long, believed, to, be, d...   
2     [A, cryptic, message, from, Bond’s, past, send...   
3     [Following, the, death, of, District, Attorney...   
4     [John, Carter, is, a, war-weary,, former, mili...   
...                                                 ...   
4804  [El, Mariachi, just, wants, to, play, his, gui...   
4805  [A, newlywed, couple's, honeymoon, is, upended...   
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   
4807  [When, ambitious, New, York, attorney, Sam, is...   
4808  [Ever, since, the, second, grade, when, he, fi...   

                                             genres  \
0     [Action, Adventure, Fantasy, Science Fiction]   
1                      [Adventure, Fantasy, Action]   
2                        [Action, Adventure, Crime]   
3                  [Action, Crime, Drama, Thriller]   
4              [Action, Adventure, Science Fiction]   
...                                             ...   
4804                      [Action, Crime, Thriller]   
4805                              [Comedy, Romance]   
4806             [Comedy, Drama, Romance, TV Movie]   
4807                                             []   
4808                                  [Documentary]   

                                               keywords  \
0     [culture clash, future, space war, space colon...   
1     [ocean, drug abuse, exotic island, east india ...   
2     [spy, based on novel, secret agent, sequel, mi...   
3     [dc comics, crime fighter, terrorist, secret i...   
4     [based on novel, mars, medallion, space travel...   
...                                                 ...   
4804  [united states–mexico barrier, legs, arms, pap...   
4805                                                 []   
4806  [date, love at first sight, narration, investi...   
4807                                                 []   
4808          [obsession, camcorder, crush, dream girl]   

                                                   cast                crew  
0     [SamWorthington, ZoeSaldana, SigourneyWeaver, ...      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley, Ste...     [GoreVerbinski]  
2     [DanielCraig, ChristophWaltz, LéaSeydoux, Ralp...         [SamMendes]  
3     [ChristianBale, MichaelCaine, GaryOldman, Anne...  [ChristopherNolan]  
4     [TaylorKitsch, LynnCollins, SamanthaMorton, Wi...     [AndrewStanton]  
...                                                 ...                 ...  
4804  [CarlosGallardo, JaimedeHoyos, PeterMarquardt,...   [RobertRodriguez]  
4805  [EdwardBurns, KerryBishé, MarshaDietlein, Cait...       [EdwardBurns]  
4806  [EricMabius, KristinBooth, CrystalLowe, GeoffG...        [ScottSmith]  
4807  [DanielHenney, ElizaCoupe, BillPaxton, AlanRuc...        [DanielHsia]  
4808  [DrewBarrymore, BrianHerzlinger, CoreyFeldman,...   [BrianHerzlinger]  

[4806 rows x 7 columns]

In [364]:
for i in ['genres','keywords']:
    def funct(x):
        l1 = []
        for j in x:
            j= j.replace(' ','')
            l1.append(j)
        return l1
    movies[i] = movies[i].apply(lambda x :funct(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/1208252187.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies[i] = movies[i].apply(lambda x :funct(x))


In [365]:
movies

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     [In, the, 22nd, century,, a, paraplegic, Marin...   
1     [Captain, Barbossa,, long, believed, to, be, d...   
2     [A, cryptic, message, from, Bond’s, past, send...   
3     [Following, the, death, of, District, Attorney...   
4     [John, Carter, is, a, war-weary,, former, mili...   
...                                                 ...   
4804  [El, Mariachi, just, wants, to, play, his, gui...   
4805  [A, newlywed, couple's, honeymoon, is, upended...   
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   
4807  [When, ambitious, New, York, attorney, Sam, is...   
4808  [Ever, since, the, second, grade, when, he, fi...   

                                            genres  \
0     [Action, Adventure, Fantasy, ScienceFiction]   
1                     [Adventure, Fantasy, Action]   
2                       [Action, Adventure, Crime]   
3                 [Action, Crime, Drama, Thriller]   
4              [Action, Adventure, ScienceFiction]   
...                                            ...   
4804                     [Action, Crime, Thriller]   
4805                             [Comedy, Romance]   
4806             [Comedy, Drama, Romance, TVMovie]   
4807                                            []   
4808                                 [Documentary]   

                                               keywords  \
0     [cultureclash, future, spacewar, spacecolony, ...   
1     [ocean, drugabuse, exoticisland, eastindiatrad...   
2     [spy, basedonnovel, secretagent, sequel, mi6, ...   
3     [dccomics, crimefighter, terrorist, secretiden...   
4     [basedonnovel, mars, medallion, spacetravel, p...   
...                                                 ...   
4804  [unitedstates–mexicobarrier, legs, arms, paper...   
4805                                                 []   
4806  [date, loveatfirstsight, narration, investigat...   
4807                                                 []   
4808           [obsession, camcorder, crush, dreamgirl]   

                                                   cast                crew  
0     [SamWorthington, ZoeSaldana, SigourneyWeaver, ...      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley, Ste...     [GoreVerbinski]  
2     [DanielCraig, ChristophWaltz, LéaSeydoux, Ralp...         [SamMendes]  
3     [ChristianBale, MichaelCaine, GaryOldman, Anne...  [ChristopherNolan]  
4     [TaylorKitsch, LynnCollins, SamanthaMorton, Wi...     [AndrewStanton]  
...                                                 ...                 ...  
4804  [CarlosGallardo, JaimedeHoyos, PeterMarquardt,...   [RobertRodriguez]  
4805  [EdwardBurns, KerryBishé, MarshaDietlein, Cait...       [EdwardBurns]  
4806  [EricMabius, KristinBooth, CrystalLowe, GeoffG...        [ScottSmith]  
4807  [DanielHenney, ElizaCoupe, BillPaxton, AlanRuc...        [DanielHsia]  
4808  [DrewBarrymore, BrianHerzlinger, CoreyFeldman,...   [BrianHerzlinger]  

[4806 rows x 7 columns]

In [366]:
movies['overview']


0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [367]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/603755214.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']


In [368]:
new_df = movies[['movie_id','title','tags']]

In [369]:
new_df['tags'] = new_df['tags'].apply(lambda x :' '.join(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/2438884289.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x :' '.join(x))


In [370]:
new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/3214958533.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())


In [371]:
new_df

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a paraplegic marine is di...  
1     captain barbossa, long believed to be dead, ha...  
2     a cryptic message from bond’s past sends him o...  
3     following the death of district attorney harve...  
4     john carter is a war-weary, former military ca...  
...                                                 ...  
4804  el mariachi just wants to play his guitar and ...  
4805  a newlywed couple's honeymoon is upended by th...  
4806  "signed, sealed, delivered" introduces a dedic...  
4807  when ambitious new york attorney sam is sent t...  
4808  ever since the second grade when he first saw ...  

[4806 rows x 3 columns]

WE WILL USE STEMMING AS WELL TO REPLACE [LOVED , LOVER , LOVE] INTO [LOVE , LOVE , LOVE]

In [372]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [373]:
ps.stem('loved') , ps.stem('loving'),ps.stem('love')

('love', 'love', 'love')

In [374]:

def stem(x):
    y = []
    for i in x.split(' '):
        y.append(ps.stem(i))
    return ' '.join(y)

In [375]:
new_df['tags'] = new_df['tags'].apply(lambda x: stem(x))

/var/folders/jy/51cp4g1n5_x_1t696_dkr24c0000gn/T/ipykernel_3604/41846828.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x: stem(x))


In [376]:
new_df

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a parapleg marin is dispa...  
1     captain barbossa, long believ to be dead, ha c...  
2     a cryptic messag from bond’ past send him on a...  
3     follow the death of district attorney harvey d...  
4     john carter is a war-weary, former militari ca...  
...                                                 ...  
4804  el mariachi just want to play hi guitar and ca...  
4805  a newlyw couple' honeymoon is upend by the arr...  
4806  "signed, sealed, delivered" introduc a dedic q...  
4807  when ambiti new york attorney sam is sent to s...  
4808  ever sinc the second grade when he first saw h...  

[4806 rows x 3 columns]

TEXT TO VECTORS CONVERSIONS

stop words wouldn't be considered in a vector

4806 movies and 5000 vectors point for each of them 

In [377]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000,stop_words='english')
vectors = cv.fit_transform(new_df['tags']).toarray()

In [378]:
vectors.shape

(4806, 5000)

In [379]:
cv.get_feature_names_out()[0:100]

array(['000', '007', '10', '100', '11', '12', '13', '14', '15', '16',
       '17', '18', '18th', '19', '1910', '1930', '1940', '1950', '1960',
       '1960s', '1970', '1970s', '1980', '1990', '19th', '19thcenturi',
       '20', '200', '20th', '24', '25', '30', '3d', '40', '50', '60',
       '70', 'aaron', 'aaroneckhart', 'aarontaylor', 'aaronyoo',
       'aasifmandvi', 'abandon', 'abduct', 'abigailbreslin', 'abil',
       'abl', 'abov', 'abus', 'academi', 'accept', 'access', 'accid',
       'accident', 'accompani', 'accomplish', 'account', 'accus', 'ace',
       'achiev', 'act', 'action', 'activ', 'activist', 'actor', 'actress',
       'actual', 'adam', 'adambrodi', 'adamgodley', 'adamgoldberg',
       'adamlefevr', 'adamsandl', 'adamscott', 'adamshankman', 'adapt',
       'add', 'addict', 'adewaleakinnuoye', 'adjust', 'admir', 'admit',
       'adolesc', 'adopt', 'ador', 'adrianmartinez', 'adrienbrodi',
       'adult', 'adulteri', 'adulthood', 'advanc', 'adventur',
       'adventure', 

In [380]:
ps.stem('actual')

'actual'

USE COSINE TO CALCULATE THE DISTANCE

In [381]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vectors)

(4806, 4806) for each movie it has find the distanbce

In [382]:
similarity

array([[1.        , 0.06668313, 0.07644708, ..., 0.04307305, 0.        ,
        0.        ],
       [0.06668313, 1.        , 0.07476672, ..., 0.02106314, 0.        ,
        0.02248469],
       [0.07644708, 0.07476672, 1.        , ..., 0.02414726, 0.        ,
        0.        ],
       ...,
       [0.04307305, 0.02106314, 0.02414726, ..., 1.        , 0.04307305,
        0.04357102],
       [0.        , 0.        , 0.        , ..., 0.04307305, 1.        ,
        0.0919601 ],
       [0.        , 0.02248469, 0.        , ..., 0.04357102, 0.0919601 ,
        1.        ]])

sorted(list(enumerate(similarity[new_df[new_df['title']=='Avatar'].index[0]])),reverse=True,key=lambda x :x[1])[1:6]

In [383]:
def recommend(movie):
    movie_index = new_df[new_df['title']==movie].index[0]
    distance = similarity[movie_index]
    movie_list  = sorted(list((enumerate(distance))),reverse=True,key= lambda x :x[1])[1:6]
    for i in movie_list:
        print(new_df.iloc[[i[0]]].title.to_numpy()[0])


In [384]:
recommend('Batman Begins')

The Dark Knight
Batman
The Dark Knight Rises
Batman
Rockaway
